File: utils/janus.py
import os
import os.path as osp
import string
import random

import asyncio
import aiohttp
import aiortc

from aiortc.contrib.media import MediaRelay
from aiortc import RTCPeerConnection

def random_trans_id():
    return "".join(random.choice(string.ascii_letters) for x in range(12))

class JanusPlugin:

    def __init__(self, session, url):
        self._queue = asyncio.Queue()
        self._session = session
        self._url = url

    async def send(self, payload):
        msg = { 'janus': 'message', 'transaction': random_trans_id() }
        msg.update(payload)
        async with self._session._http.post(self._url, json=msg) as response:
            data = await response.json()
            assert data['janus'] == 'ack'
        # Wait for the asynchronized response. The polling task in the session
        # will put the response in the queue of the plugin.
        response = await self._queue.get()
        assert response['transaction'] == msg['transaction']
        return response


class JanusSession:

    def __init__(self, url):
        # Janus Information
        self._url = url
        self._http = aiohttp.ClientSession()
        self._session_url = None
        self._poll_task = None
        self._plugins = {}
        # AIORTC Information
        self._pcs = {}
        self._relays = {}
        self._tracks = {}

    def createRTCPeerConnection(self, name):
        pc = RTCPeerConnection()
        self._pcs[name] = pc
        return pc

    def getRTCPeerConnection(self, name):
        return self._pcs.get(name, None)

    def createMediaRelay(self, name):
        relay = MediaRelay()
        self._relays[name] = relay
        return relay

    def getMediaRelay(self, name):
        return self._relays.get(name, None)

    def saveTrack(self, name, track):
        self._tracks[name] = track

    def accessTrack(self, name):
        return self._tracks.get(name, None)

    async def attach(self, plugin_name: str) -> JanusPlugin:
        msg = {
            "janus": "attach",
            "plugin": plugin_name,
            "transaction": random_trans_id()
        }
        async with self._http.post(self._session_url, json=msg) as response:
            data = await response.json()
            assert data['janus'] == 'success'
            plugin_id = data['data']['id']
            plugin = JanusPlugin(self, osp.join(self._session_url, str(plugin_id)))
            self._plugins[plugin_id] = plugin
            return plugin

    async def create(self):
        """Get the available session ID to be used"""
        msg = {
            "janus": "create",
            "transaction": random_trans_id()
        }
        async with self._http.post(self._url, json=msg) as response:
            data = await response.json()
            assert data['janus'] == 'success'
            session_id = data['data']['id']
            self._session_url = osp.join(self._url, str(session_id))

        self._poll_task = asyncio.ensure_future(self._poll())

    async def destroy(self):
        if self._poll_task:
            self._poll_task.cancel()
            self._poll_task = None

        if self._session_url:
            msg = { 'janus': 'destroy', 'transaction': random_trans_id() }
            async with self._http.post(self._session_url, json=msg) as response:
                data = await response.json()
                assert data['janus'] == 'success'
            self._session_url = None

        if self._http:
            await self._http.close()
            self._http = None

        for name, pc in self._pcs.items():
            await pc.close()

    async def _poll(self):
        while True:
            params = { "maxev": 1 }
            async with self._http.get(self._session_url, params=params) as response:
                data = await response.json()
                if data['janus'] == 'event':
                    plugin = self._plugins.get(data['sender'], None)
                    if plugin:
                        await plugin._queue.put(data)

File: js/janus.js
// Global variables
var eventQueue = [];
var pluginQueues = {};
var pluginID2Name = {};

// Exported functions
function getPluginQueue(name) {
  return pluginQueues[name];
}

function getPluginName(id) {
  return pluginID2Name[id];
}

// Helper functions
function _janus_msleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

function _janus_random_trans_id(){
  const chars ='ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';

  let result = [];
  for ( let i = 0; i < 12; i++ ) {
    result += chars.charAt(
      Math.floor(Math.random() * chars.length)
    );
  }

  return result;
}

function _janus_join_url(base, token) {
  if (base.substr(-1) != '/')
    let url = base + '/';
  else
    let url = base;

  return url + String(token);
}

function _janus_join_url_params(base, params) {
  if (base.substr(-1) != '/')
    let url = base + '/';
  else
    let url = base;

  let count = 0;
  for (const [key, value] of Object.entries(params)) {
    if (count == 0)
      url += `?${key}=${value}`;
    else
      url += `&${key}=${value}`;
    count += 1;

  return url;
}

// Exported Asynchronized Functions
async function createJanusSession(janusUrl) {
  const msg = {
    'janus': 'create',
    'transaction': _janus_random_trans_id(),
  }

  // Create janus session handle
  let response = await fetch(janusUrl, {
    method: 'POST',
    body: JSON.stringify(msg)
  });
  let data = await response.json();

  // Response Check
  if (data['janus'] != 'success') {
    throw new Error("Fail to create janus session");
  }

  // Return janus session url
  const sessionID = data['data']['id']
  const sessionUrl = _janus_join_url(janusUrl, sessionID);

  return sessionUrl
}

async function attachJanusPlugin(sessionUrl, pluginName) {
  const msg = {
    janus: 'attach',
    plugin: pluginName,
    transaction: _janus_random_trans_id()
  };

  // Craet janus plugin handle
  let response = await fetch(sessionUrl, {
    method: 'POST',
    body: JSON.stringify(msg)
  });
  let data = await response.json();

  // Response Check
  if (data['janus'] != 'success') {
    throw new Error("Fail to attach to plugin "+pluginName);
  }

  // Extract plugin handle
  const pluginID = data['data']['id'];
  const pluginUrl = _janus_join_url(sessionUrl, pluginID);

  // Create dedicated event queue for each plugin handle
  pluginQueues[pluginID] = [];
  pluginID2Name[pluginID] = pluginName;

  return pluginUrl;
}

async function sendPluginMessage(pluginUrl, data) {
  let msg = {
    janus: 'message',
    transaction: _janus_random_trans_id(),
  };

  // Update message with plugin specific data
  for (const [key, value] of Object.entries(data)) {
    msg[key] = value;

  // Send Message to specified plugin handle
  let response = await fetch(pluginUrl, {
    method: 'POST',
    body: JSON.stringify(msg)
  });
  let data = await response.json();

  // Response check
  if (data['janus'] != 'ack') {
    throw new Error("Plugin does not ack your request");
  }

  // Wait for polled event data matched with the requested transaction ID
  let elapsed = 0;
  while (elapsed < 500) {
    // Check for recently got event data
    let event = eventQueue.shift();
    if (
      'transaction' in event
      && event.transaction == msg.transaction
    )
      return event;

    // Push back the event data into the event queue
    eventQueue.push(event);
    await _janus_msleep(100);
    elapsed += 100;
  }

  throw new Error("Fail to get responed event data within 500 msec");
}

async function janusEventSubscriber(sessionUrl) {

  const params = { maxev: '1' };
  const pollUrl = _janus_join_url_params(sessionUrl, params);

  response = await fetch(pollUrl, { method: 'GET' });
  data = await response.json();

  if (data['janus'] == 'event') {
    if (data['sender'] in pluginQueues) {
      queue = pluginQueues[String(data['sender'])];
      queue.push(data);
      eventQueue.push(data);
    }
  }

}


async function createMediaClient(pluginUrl, media) {
  const config = {
    sdpSemantics: 'unified-plan',
    iceServers: [{urls: ['stun:stun.l.google.com:19302']}]
  };

  let pc = new RTCPeerConnection(config);

  // Attach event listener for available tracks (audio/video)
  pc.addEventListener('track', function(evt) {
    if (evt.track.kind == 'video') {
      console.log(evt);
      let video = document.getElementById(`${media.id}_video`);
      console.log(video);
      video.srcObject = evt.streams[0];
    } else {
      console.log(evt);
      let audio = document.getElementById(`${media.id}_audio`);
      audio.srcObject = evt.streams[0];
    }
  });

  pc.addTransceiver('video', {direction: 'recvonly'});
  pc.addTransceiver('audio', {direction: 'recvonly'});

  // Request for SDP information of the publisher (feed)
  body = {
    ptype: 'subscriber',
    request: 'join',
    room: roomID,
    feed: pub.id,
  };
  data = await sendPluginMessage(pluginUrl, body);
  description = new RTCSessionDescription(data['jsep']);

  await pc.setRemoteDescription(description);
  await pc.setLocalDescription(await pc.createAnswer())

  console.log(pc.localDescription);
  body = { request: 'start' }
  jsep = { sdp: pc.localDescription.sdp, type: pc.localDescription.type };
  data = await sendPluginMessage(pluginUrl, body, jsep);

  console.log(data);
}

function createMediaComponent(id, name) {
  let div = document.createElement('div');
  div.setAttribute('id', id);

  let h2 = document.createElement('h2');
  h2.setAttribute('id', `${id}_name`);
  h2.innerHTML = name;

  let audio = document.createElement('audio');
  audio.setAttribute('id', `${id}_audio`);
  audio.setAttribute('autoplay', 'true');

  let video = document.createElement('video');
  video.setAttribute('id', `${id}_video`);
  video.setAttribute('autoplay', 'true');
  video.setAttribute('playsinline', 'true');

  div.appendChild(h2);
  div.appendChild(audio);
  div.appendChild(video);

  return div;
}
